### PR TITLE
Adds a kubeadm config images pull command

### LINKED ISSUE #812: Add a command to pull necessary images
## Is this a BUG REPORT or FEATURE REQUEST?
FEATURE REQUEST

We should add a feature around `kubeadm config list-images` that also pulls the images using `crictl`. This way we don't have to write special code for every container runtime, but we will have to special case docker if this is run before the kubelet is running (and thus hosting the docker CRI shim (shim might be the wrong word))

### PR DESCRIPTION
**Pull request by chuckha**
This command will use crictl or docker to pull images locally.

The dockerfall back is needed because in some cases the kubelet is not
yet running so there is no CRI dockershim socket available.

Fixes kubernetes/kubeadm#812

Signed-off-by: Chuck Ha <ha.chuck@gmail.com>

**What this PR does / why we need it**:
This PR adds a command to the `kubeadm config images` subcommand to pull images using either crictl or docker.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#812

**Special notes for your reviewer**:

**Release note**:
```release-note
Adds a `kubeadm config images pull` command to pull container images used by kubeadm.
```

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

File: cmd/kubeadm/app/images/puller.go
Comment (timothysc) [inline cmd/kubeadm/app/images/puller.go]: Don't we need todo version magic checking to ensure that the crictl can pull for the runtimes we are endorsing?  We'll need to ship a version with kubeadm that we we vet.

File: cmd/kubeadm/app/cmd/config.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/config.go]: did these two PRs overlap?
https://github.com/kubernetes/kubernetes/pull/63811

oddly, for the other one the bots are saying `PR is unable to be automatically merged. Needs rebase.` but github thinks there is no need for rebase.

the refactor and `pull` addition could have been separate commits.

File: cmd/kubeadm/app/cmd/config.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/config.go]: it's OK to remove the `( )` here.

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@chuckha 
I suggested some refactoring for a better separation between UX and the image pull machinery.
However the PR overall is fine for me. Thanks!

File: cmd/kubeadm/app/cmd/config.go
Comment (chuckha) [reply cmd/kubeadm/app/cmd/config.go]: Yeah, that was me basing this branch off that one. I forgot to mention that, my mistake. The refactor and the addition are in two separate commits. I just fixed the merge conflict, it should look a lot better now.

File: cmd/kubeadm/app/cmd/config.go
Comment (chuckha) [reply cmd/kubeadm/app/cmd/config.go]: I was thinking about this as if it were an internal interface (which I should have then named it `puller` instead) -- not really something for reuse, but I don't see any reason why it can't live in the images package.

The `ImagesPull` was not designed for use elsewhere, but looking at it a bit closer, the only thing I can see that is specific to this command is the `glog` call.

Would you be ok if I moved the Puller interface out to the `images` package but kept ImagesPull internal to this package for now?

File: cmd/kubeadm/app/images/puller.go
Comment (chuckha) [reply cmd/kubeadm/app/images/puller.go]: Why make it private? I'd also argue that the name should not reveal the implementation as this uses either docker or crictl to pull images

File: cmd/kubeadm/app/images/puller.go
Comment (chuckha) [reply cmd/kubeadm/app/images/puller.go]: You are 💯% correct. We'll be addressing that in this issue https://github.com/kubernetes/kubeadm/issues/811

File: cmd/kubeadm/app/images/puller.go
Comment (chuckha) [inline cmd/kubeadm/app/images/puller.go]: This logic is wrong. We want to use docker if the cri socket is the default docker socket. Otherwise, use crictl

**Review by timothysc — APPROVED** [review]
/approve 

@fabriziopandini please lgtm when you are happy with the change.

File: cmd/kubeadm/app/cmd/config.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/config.go]: Ever if there is no real use case for reuse now, I would like to have the capability to pull a set of images implemented in the `app/images` package instead of in the UX

> ..., the only thing I can see that is specific to this command is the `glog` call.
I don't see this as a constraint, e.g. there are already several glog calls in phases

File: cmd/kubeadm/app/images/puller.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/images/puller.go]: The initial lower case was a typo, sorry 

However I'm still not fully comfortable about naming of `ImagePull` and `ImagePuller` because those confused me while reading the code. If we are keeping  'ImagePuller' because it's implementation, what about `ImagePull` ->  `ImageSetPuller` or something that makes more clear it's responsibility to coordinate the pull process for a set of images

File: cmd/kubeadm/app/cmd/config.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/config.go]: must use `&` pointer here

File: cmd/kubeadm/app/cmd/config.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/config.go]: `kubeadmscheme.Scheme.Default(cfg)`
where `kubeadmscheme "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/scheme"`

File: cmd/kubeadm/app/cmd/config.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/config.go]: nit: maybe use `%q` instead of `[%v]`?

File: cmd/kubeadm/app/cmd/config.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/config.go]: just `[images]` is fine for me here. As `image` is a string, use `%s\n` (don't forget the newline)

File: cmd/kubeadm/app/cmd/config.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/config.go]: Why change this?

File: cmd/kubeadm/app/images/puller.go
Comment (luxas) [inline cmd/kubeadm/app/images/puller.go]: `kubeadmapiv1alpha1`

File: cmd/kubeadm/app/images/puller.go
Comment (luxas) [inline cmd/kubeadm/app/images/puller.go]: What happens if `crictlPath` is "" but there isn't an error?

**Review by luxas — COMMENTED** [review]
A couple of comments that need fixing, but overall this LGTM :+1:. Thanks for this patch!

File: cmd/kubeadm/app/cmd/config.go
Comment (chuckha) [reply cmd/kubeadm/app/cmd/config.go]: sure, seems reasonable. I need to understand this stuff better

File: cmd/kubeadm/app/cmd/config.go
Comment (luxas) [reply cmd/kubeadm/app/cmd/config.go]: It's _technically_ not a must, but practically. If you init it as a pointer, you don't have to remember to pass `&` all the time. I'm pretty sure most things break if someone somehow forgets to pass the pointer at some time. Would be super hard to debug and notice.

Comment (chuckha) [reply cmd/kubeadm/app/cmd/config.go]: fair point 👍 thanks for explaining what you meant

File: cmd/kubeadm/app/images/puller.go
Comment (neolit123) [reply cmd/kubeadm/app/images/puller.go]: `&&` -> `||`

Comment (chuckha) [reply cmd/kubeadm/app/images/puller.go]: Also wrong logic from before -- If we *need* crictl (our crisocket is not the default) then we fail. Otherwise it's fine that we don't have crictl available (since we use docker instead).

**Review by luxas — APPROVED** [review]
/lgtm

Thanks a lot Chuck :tada:!

**Pull request merged**
